### PR TITLE
[security] Adds SecurityUtils

### PR DESCRIPTION
**Pull request by sdelamo**
Close: https://github.com/micronaut-projects/micronaut-core/issues/558

### CODE REVIEW

**Comment by sdelamo** [comment]
@jameskleeh  and @graemerocher  I would like to comment this with you. 

I have changed `@Secured` to be a AOP Around advice so that `@Secured` can be used with things such as Web Sockets etc.

It is possible to use it now to secure an method in a Service as illustrated by this test:

https://github.com/micronaut-projects/micronaut-core/blob/df15ea8d62bdd13503c960323c6ad2555dcdcf8a/security/src/test/groovy/io/micronaut/security/atsecured/AtSecuredAppliedToServiceMethodSpec.groovy#L65

However, I bypass its use in controllers: 

https://github.com/micronaut-projects/micronaut-core/blob/df15ea8d62bdd13503c960323c6ad2555dcdcf8a/security/src/main/java/io/micronaut/security/SecuredInterceptor.java#L52

> If {@link io.micronaut.security.Secured} annotation is present in a class annotated with {@link io.micronaut.http.annotation.Controller}, the interceptor does nothing because it is handled by {@link io.micronaut.security.rules.SecuredAnnotationRule}.

Skipping its use in @Controller and leaving that to `SecuredAnnoatationRule` allows things such as `@Secure` override with intercept urls maps. See test: 

https://github.com/micronaut-projects/micronaut-core/blob/df15ea8d62bdd13503c960323c6ad2555dcdcf8a/security/src/test/groovy/io/micronaut/security/overrideatsecured/OverrideAtSecuredControllerSpec.groovy#L44

**Comment by sdelamo** [comment]
>> "is possible to use it now to secure an method in a Service" - 
> What use case is being solved with this change?

It will allow to put security checks downstream not just in a Controller method.  

>> "allows things such as @secure override with intercept urls maps" - I'm not sure that is a good idea.

I think it will give you flexibility in case you embeed a dependency which includes a controller which uses @Secured in a method and you want to teak to fit your security needs.

File: security/src/main/java/io/micronaut/security/SecuredInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+/**
+ * Interceptor implementation for the {@link io.micronaut.security.Secured} annotation.
+ *
+ * If {@link io.micronaut.security.Secured} annotation is present in a class annotated with {@link io.micronaut.http.annotation.Controller}, the interceptor does nothing because it is handled by {@link io.micronaut.security.rules.SecuredAnnotationRule}.
+ *
+ * @author Sergio del Amo
+ * @since 1.0
+ */
+@Singleton
+public class SecuredInterceptor implements MethodInterceptor<Object, Object> {
```
Comment (graemerocher) [inline security/src/main/java/io/micronaut/security/SecuredInterceptor.java]: Needs to define an order. See `InterceptPhase`.

**Comment by sdelamo** [comment]
This will not be merged. I will create separated PRs for pieces of functionality.

**Pull request closed without merging**